### ISSUE TITLE
Referencing file where new es.Client causes 'module not found' error.

### ISSUE THREAD

**Issue by jamiedumont** [elastic/elasticsearch-js#405]
Hey all,

I've been struggling with a problem for far too long. It's been troubling me off and on for a few weeks, and I can't seem to find the solution. I'm not sure whether it's Meteor's build system or ES, but I only seem to be getting this issue with this one file where the new ES Client is created.

I'm trying to import functions into my `/imports/api/*/methods.js` files. It works absolutely fine in one instance, however, as soon as I try to import the same function elsewhere (another file); I get the following error from my browser console.

`Uncaught Error: Cannot find module '/imports/api/elastic/server/elastic.js'`

The functions are being created and exported from `/imports/api/elastic/server/elastic.js`, however I've tried this file locating elsewhere (server side) with no effect. Here is that file for reference...

    import es from 'elasticsearch';
    import { Meteor } from 'meteor/meteor';
    import { Promise } from 'meteor/promise';

    const es = new Elastic.Client({
     host: [{
       host: 'hidden-cluster-address-here',
       auth: Meteor.settings.ESCloudRW,
     },],
    });

    export function esIndexExists(indexName) {
     const promise = es.indices.exists({ index: indexName });
     return Promise.await(promise);
    }

    export function esCreateIndex(indexName) {
     const promise = es.indices.create({ index: indexName });
     return Promise.await(promise);
    }

I know the ES client needs to be defined on the server, and I've been sure to make certain it is.

The error is thrown not by attempting to use the function multiple times, but by importing it. Maybe I'm encountering this error because I can't reference the file more than once due to the presence of the ES client in that file?

Importing *any* of those functions from the `elastic.js` file once is fine, additional imports in other files (even if different modules) causes the error.

Are there any known issues surrounding referencing the same file more than once?

I'm all out of ideas. Help!

Cheers,

Jamie.

Sidenote: Locating all the methods that require these ES functions in one file stops the error, but methods that previously ran, then cause incomprehensible errors in the terminal.

**Comment by spalger** [comment]
Hmm, there shouldn't be any issues related to requiring the file more than once. In node `require()` is idempotent, call it 1 or 100 times and the result will be the same. I'm 99.9% this has to do with Meteor. Can you share the stacktrace you get when you see `Uncaught Error: Cannot find module '/imports/api/elastic/server/elastic.js'`? Is this happening in the browser?

**Comment by jamiedumont** [comment]
I had thought Meteor may be the culprit, but I haven't discovered any similar issues with my numerous other imports. 

I'm in the process of swapping the Meteor build tool out for Webpack to see if that resolves the issue, and identifies the source of the problem. 

In the meantime, I've attached the stack trace; and yes, this is happening in the browser.
![something](https://cloud.githubusercontent.com/assets/5637895/15371594/2797a300-1d34-11e6-843b-3e17369da205.jpg)

**Comment by jamiedumont** [comment]
Right, after way too much swearing and messing about; I finally found the issue...Meteor.

More precisely, Meteor's client/server-side nature makes it difficult to be certain where code runs. My issue was that the client was trying to `import` a file that can only exist server-side (due to ES). That caused the build to fail big, rather than silently client-side whilst working server-side.

I resolved it by defining my method inside a shared file, but specifying that it only be run on the server. Unfortunately I can't use `import` so I have to fall back to using `require`. The code now looks like this:

    import { Meteor } from 'meteor/meteor';
    if (Meteor.isServer) {

      let client = require('/imports/api/elastic/server/client.js');

      Meteor.methods({
        addBike(bikeAttributes) {
          check(bikeAttributes, Object);
    
          // lots of other code here....

          // Insert in ES cluster
          client.esAddDocument('marketplace', 'bikes', bike._id, bike);
        },
      });
    }

`esAddDocument()` is a function that I defined in the server-only file above (enforced by living within a `/server` folder. Meteor's black box magic can be quite tiresome sometimes...
